torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -757,7 +757,6 @@ def all_gather_inputs(self) -> List[torch.Tensor]:  # 1D
 
     @property
     def unsharded_param(self) -> nn.Parameter:  # ND
-        self._assert_in_states(ShardedState.UNSHARDED)
```

Comment:
The main change here is that we used to define `unsharded_param` as a property so that we can assert that it is only accessed when in the unsharded state. However, with this new API, it is possible to get grads on the unsharded parameter without being in the unsharded state (since the unsharded parameter data is not allocated).

We may refactor in the future to get rid of `unsharded_param` as a property then.